File: src/main/summaryService.ts
import { existsSync, readFileSync, writeFileSync } from 'node:fs';
import { join } from 'node:path';
import { spawn } from 'node:child_process';
import { getNoteContentHash, getSummaryCacheKey } from './obsidianScanner';
import { getSummaryCacheDir, type AppConfig } from './config';
import { formatSummaryBullets } from './summaryFormat';

const SUMMARY_TIMEOUT_MS = 120_000;
const MAX_LLM_RELATED_NOTES = 8;

export interface LlmRelatedNote {
  title: string;
}

export interface NoteSummaryResult {
  summary: string;
  relatedNotes: LlmRelatedNote[];
}

interface CachedNoteSummary {
  contentHash: string;
  summary: string;
  relatedNotes: LlmRelatedNote[];
}

function buildPrompt(title: string, content: string): string {
  const trimmedContent = content.slice(0, 12_000);
  return [
    'Return a single JSON object and nothing else — no preamble, no explanation, no markdown fences.',
    '{"summary":"- First short point\\n- Second short point\\n- Third short point","relatedNotes":[{"title":"..."}]}',
    '',
    'Rules:',
    '- summary: Markdown bullet list ONLY. Every line must start with "- ". Use 2-5 bullets. Keep each bullet under 12 words.',
    `- relatedNotes: up to ${MAX_LLM_RELATED_NOTES} notes linked via [[wiki links]] in the note or clearly related. Use exact note titles from the note body.`,
    '- If none apply, return "relatedNotes":[]',
    '- Do not mention permissions, vault access, or that you cannot read files.',
    '',
    `Note title: ${title}`,
    '',
    trimmedContent,
  ].join('\n');
}

function readCachedSummary(
  cacheDir: string,
  cacheKey: string,
  contentHash: string,
): CachedNoteSummary | null {
  const cachePath = join(cacheDir, `${cacheKey}.json`);
  if (!existsSync(cachePath)) {
    return null;
  }

  try {
    const parsed = JSON.parse(readFileSync(cachePath, 'utf8')) as CachedNoteSummary;
    if (!parsed.summary || parsed.contentHash !== contentHash) {
      return null;
    }

    return normalizeCachedSummary({
      contentHash: parsed.contentHash,
      summary: parsed.summary.trim(),
      relatedNotes: Array.isArray(parsed.relatedNotes) ? parsed.relatedNotes : [],
    });
  } catch {
    return null;
  }
}

function writeCachedSummary(
  cacheDir: string,
  cacheKey: string,
  result: CachedNoteSummary,
): void {
  const cachePath = join(cacheDir, `${cacheKey}.json`);
  writeFileSync(cachePath, JSON.stringify(result, null, 2), 'utf8');
}

function findEmbeddedJsonObject(text: string): string | null {
  const start = text.indexOf('{');
  if (start === -1) {
    return null;
  }

  let depth = 0;
  let inString = false;
  let escaped = false;

  for (let index = start; index < text.length; index += 1) {
    const char = text[index];

    if (escaped) {
      escaped = false;
      continue;
    }

    if (char === '\\' && inString) {
      escaped = true;
      continue;
    }

    if (char === '"') {
      inString = !inString;
      continue;
    }

    if (inString) {
      continue;
    }

    if (char === '{') {
      depth += 1;
    }

    if (char === '}') {
      depth -= 1;
      if (depth === 0) {
        return text.slice(start, index + 1);
      }
    }
  }

  return null;
}

function extractJsonPayload(raw: string): string {
  const trimmed = raw.trim();
  const fencedMatch = trimmed.match(/```(?:json)?\s*([\s\S]*?)\s*```/i);
  if (fencedMatch?.[1]) {
    return fencedMatch[1].trim();
  }

  const embedded = findEmbeddedJsonObject(trimmed);
  if (embedded) {
    return embedded;
  }

  return trimmed;
}

function normalizeRelatedNotes(value: unknown): LlmRelatedNote[] {
  if (!Array.isArray(value)) {
    return [];
  }

  return value
    .filter((item): item is LlmRelatedNote => {
      return Boolean(item) && typeof item === 'object' && typeof item.title === 'string';
    })
    .map((item) => ({ title: item.title.trim() }))
    .filter((item) => item.title.length > 0)
    .slice(0, MAX_LLM_RELATED_NOTES);
}

function tryParseSummaryJson(jsonText: string): NoteSummaryResult | null {
  try {
    const parsed = JSON.parse(jsonText) as Partial<NoteSummaryResult>;
    const summary = typeof parsed.summary === 'string' ? parsed.summary.trim() : '';

    if (!summary) {
      return null;
    }

    return {
      summary: formatSummaryBullets(summary),
      relatedNotes: normalizeRelatedNotes(parsed.relatedNotes),
    };
  } catch {
    return null;
  }
}

function looksLikeMalformedSummary(summary: string): boolean {
  const trimmed = summary.trim();
  return trimmed.startsWith('{') || trimmed.includes('"summary":') || trimmed.includes('"relatedNotes"');
}

function parseSummaryResponse(raw: string): NoteSummaryResult {
  const candidates = new Set<string>();

  candidates.add(extractJsonPayload(raw));

  const embedded = findEmbeddedJsonObject(raw);
  if (embedded) {
    candidates.add(embedded);
  }

  for (const candidate of candidates) {
    const parsed = tryParseSummaryJson(candidate);
    if (parsed) {
      return parsed;
    }
  }

  const stripped = raw.replace(/\{[\s\S]*\}/, '').trim();
  if (stripped) {
    return {
      summary: formatSummaryBullets(stripped),
      relatedNotes: [],
    };
  }

  return {
    summary: 'Summary could not be parsed. Use Regenerate summary in settings.',
    relatedNotes: [],
  };
}

function normalizeCachedSummary(cached: CachedNoteSummary): CachedNoteSummary {
  if (!looksLikeMalformedSummary(cached.summary)) {
    return cached;
  }

  const repaired = parseSummaryResponse(cached.summary);
  return {
    contentHash: cached.contentHash,
    summary: repaired.summary,
    relatedNotes: repaired.relatedNotes.length > 0 ? repaired.relatedNotes : cached.relatedNotes,
  };
}

function runClaude(config: AppConfig, prompt: string): Promise<string> {
  return new Promise((resolve, reject) => {
    const args = [
      '-p',
      '--no-session-persistence',
      '--permission-mode',
      'dontAsk',
      '--output-format',
      'text',
    ];

    const child = spawn(config.claudeBinary, args, {
      env: {
        ...process.env,
        CC_HEADLESS: '1',
      },
      stdio: ['pipe', 'pipe', 'pipe'],
    });

    child.stdin.write(prompt);
    child.stdin.end();

    let stdout = '';
    let stderr = '';

    const timeout = setTimeout(() => {
      child.kill('SIGTERM');
      reject(new Error('Claude CLI timed out after 120 seconds'));
    }, SUMMARY_TIMEOUT_MS);

    child.stdout.on('data', (chunk: Buffer) => {
      stdout += chunk.toString();
    });

    child.stderr.on('data', (chunk: Buffer) => {
      stderr += chunk.toString();
    });

    child.on('error', (error) => {
      clearTimeout(timeout);
      reject(new Error(`Failed to run Claude CLI (${config.claudeBinary}): ${error.message}`));
    });

    child.on('close', (code) => {
      clearTimeout(timeout);
      const output = stdout.trim();

      if (code !== 0) {
        reject(new Error(stderr.trim() || `Claude CLI exited with code ${code}`));
        return;
      }

      if (!output) {
        reject(
          new Error(
            'Claude CLI returned empty output. Run `claude -p "hello"` in Terminal to verify your CLI setup.',
          ),
        );
        return;
      }

      resolve(output);
    });
  });
}

export async function summarizeNote(
  config: AppConfig,
  title: string,
  content: string,
  filePath: string,
  options: { bypassCache?: boolean } = {},
): Promise<NoteSummaryResult> {
  const cacheDir = getSummaryCacheDir(config);
  const contentHash = getNoteContentHash(content);
  const cacheKey = getSummaryCacheKey(filePath);

  if (!options.bypassCache) {
    const cached = readCachedSummary(cacheDir, cacheKey, contentHash);
    if (cached) {
      return {
        summary: formatSummaryBullets(cached.summary),
        relatedNotes: cached.relatedNotes,
      };
    }
  }

  const rawResponse = await runClaude(config, buildPrompt(title, content));
  const result = parseSummaryResponse(rawResponse);
  writeCachedSummary(cacheDir, cacheKey, {
    contentHash,
    summary: result.summary,
    relatedNotes: result.relatedNotes,
  });
  return result;
}

File: macos/Shared/Services/SummaryService.swift
import Foundation

enum SummaryService {
    private static let summaryTimeoutSeconds: TimeInterval = 120
    private static let maxLlmRelatedNotes = 8

    private struct CachedNoteSummary: Codable {
        let contentHash: String
        let summary: String
        let relatedNotes: [LlmRelatedNote]
    }

    static func summarizeNote(
        config: AppConfig,
        title: String,
        content: String,
        filePath: String,
        bypassCache: Bool = false
    ) async throws -> NoteSummaryResult {
        let cacheDir = ConfigStore.summaryCacheDirectory(for: config)
        let contentHash = VaultScanner.contentHash(for: content)
        let cacheKey = VaultScanner.summaryCacheKey(for: filePath)

        if !bypassCache, let cached = readCachedSummary(cacheDir: cacheDir, cacheKey: cacheKey, contentHash: contentHash) {
            return NoteSummaryResult(
                summary: SummaryFormat.formatSummaryBullets(cached.summary),
                relatedNotes: cached.relatedNotes
            )
        }

        let rawResponse = try await runClaude(config: config, prompt: buildPrompt(title: title, content: content))
        let result = parseSummaryResponse(rawResponse)
        writeCachedSummary(
            cacheDir: cacheDir,
            cacheKey: cacheKey,
            result: CachedNoteSummary(
                contentHash: contentHash,
                summary: result.summary,
                relatedNotes: result.relatedNotes
            )
        )
        return result
    }

    private static func buildPrompt(title: String, content: String) -> String {
        let trimmedContent = String(content.prefix(12_000))
        return [
            "Return a single JSON object and nothing else — no preamble, no explanation, no markdown fences.",
            #"{"summary":"- First short point\n- Second short point\n- Third short point","relatedNotes":[{"title":"..."}]}"#,
            "",
            "Rules:",
            "- summary: Markdown bullet list ONLY. Every line must start with \"- \". Use 2-5 bullets. Keep each bullet under 12 words.",
            "- relatedNotes: up to \(maxLlmRelatedNotes) notes linked via [[wiki links]] in the note or clearly related. Use exact note titles from the note body.",
            "- If none apply, return \"relatedNotes\":[]",
            "- Do not mention permissions, vault access, or that you cannot read files.",
            "",
            "Note title: \(title)",
            "",
            trimmedContent,
        ].joined(separator: "\n")
    }

    private static func readCachedSummary(cacheDir: URL, cacheKey: String, contentHash: String) -> CachedNoteSummary? {
        let cachePath = cacheDir.appendingPathComponent("\(cacheKey).json")
        guard FileManager.default.fileExists(atPath: cachePath.path),
              let data = try? Data(contentsOf: cachePath),
              let parsed = try? JSONDecoder().decode(CachedNoteSummary.self, from: data),
              !parsed.summary.isEmpty,
              parsed.contentHash == contentHash else {
            return nil
        }

        return normalizeCachedSummary(parsed)
    }

    private static func writeCachedSummary(cacheDir: URL, cacheKey: String, result: CachedNoteSummary) {
        let cachePath = cacheDir.appendingPathComponent("\(cacheKey).json")
        if let data = try? JSONEncoder().encode(result) {
            try? data.write(to: cachePath)
        }
    }

    private static func normalizeCachedSummary(_ cached: CachedNoteSummary) -> CachedNoteSummary {
        if !looksLikeMalformedSummary(cached.summary) {
            return cached
        }

        let repaired = parseSummaryResponse(cached.summary)
        return CachedNoteSummary(
            contentHash: cached.contentHash,
            summary: repaired.summary,
            relatedNotes: repaired.relatedNotes.isEmpty ? cached.relatedNotes : repaired.relatedNotes
        )
    }

    private static func looksLikeMalformedSummary(_ summary: String) -> Bool {
        let trimmed = summary.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.hasPrefix("{") || trimmed.contains("\"summary\":") || trimmed.contains("\"relatedNotes\":")
    }

    private static func parseSummaryResponse(_ raw: String) -> NoteSummaryResult {
        var candidates = Set<String>()
        candidates.insert(extractJsonPayload(raw))

        if let embedded = findEmbeddedJsonObject(raw) {
            candidates.insert(embedded)
        }

        for candidate in candidates {
            if let parsed = tryParseSummaryJson(candidate) {
                return parsed
            }
        }

        let stripped = raw.replacingOccurrences(of: "\\{[\\s\\S]*\\}", with: "", options: .regularExpression)
            .trimmingCharacters(in: .whitespacesAndNewlines)

        if !stripped.isEmpty {
            return NoteSummaryResult(
                summary: SummaryFormat.formatSummaryBullets(stripped),
                relatedNotes: []
            )
        }

        return NoteSummaryResult(
            summary: "Summary could not be parsed. Use Regenerate summary in settings.",
            relatedNotes: []
        )
    }

    private static func tryParseSummaryJson(_ jsonText: String) -> NoteSummaryResult? {
        guard let data = jsonText.data(using: .utf8),
              let parsed = try? JSONDecoder().decode(PartialSummaryResponse.self, from: data),
              let summary = parsed.summary?.trimmingCharacters(in: .whitespacesAndNewlines),
              !summary.isEmpty else {
            return nil
        }

        return NoteSummaryResult(
            summary: SummaryFormat.formatSummaryBullets(summary),
            relatedNotes: normalizeRelatedNotes(parsed.relatedNotes)
        )
    }

    private struct PartialSummaryResponse: Decodable {
        let summary: String?
        let relatedNotes: [LlmRelatedNote]?
    }

    private static func normalizeRelatedNotes(_ value: [LlmRelatedNote]?) -> [LlmRelatedNote] {
        guard let value else { return [] }
        return value
            .map { LlmRelatedNote(title: $0.title.trimmingCharacters(in: .whitespacesAndNewlines)) }
            .filter { !$0.title.isEmpty }
            .prefix(maxLlmRelatedNotes)
            .map { $0 }
    }

    private static func extractJsonPayload(_ raw: String) -> String {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if let range = trimmed.range(of: #"```(?:json)?\s*([\s\S]*?)\s*```"#, options: .regularExpression) {
            let match = String(trimmed[range])
            let inner = match
                .replacingOccurrences(of: "```json", with: "")
                .replacingOccurrences(of: "```", with: "")
                .trimmingCharacters(in: .whitespacesAndNewlines)
            if !inner.isEmpty { return inner }
        }

        if let embedded = findEmbeddedJsonObject(trimmed) {
            return embedded
        }

        return trimmed
    }

    private static func findEmbeddedJsonObject(_ text: String) -> String? {
        guard let start = text.firstIndex(of: "{") else { return nil }

        var depth = 0
        var inString = false
        var escaped = false

        var index = start
        while index < text.endIndex {
            let char = text[index]

            if escaped {
                escaped = false
            } else if char == "\\" && inString {
                escaped = true
            } else if char == "\"" {
                inString.toggle()
            } else if !inString {
                if char == "{" { depth += 1 }
                if char == "}" {
                    depth -= 1
                    if depth == 0 {
                        return String(text[start...index])
                    }
                }
            }

            index = text.index(after: index)
        }

        return nil
    }

    private static func runClaude(config: AppConfig, prompt: String) async throws -> String {
        guard let executableURL = ClaudeBinaryResolver.resolve(configuredBinary: config.claudeBinary) else {
            throw SummaryServiceError.cliFailed(
                "Claude CLI not found. Install it or set claudeBinary in config.json to its full path (e.g. /opt/homebrew/bin/claude)."
            )
        }

        // #region agent log
        DebugSessionLog.write(
            hypothesisId: "H-claude",
            location: "SummaryService.swift:runClaude",
            message: "Resolved Claude binary",
            data: ["path": executableURL.path],
            runId: "claude-fix"
        )
        // #endregion

        return try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<String, Error>) in
            let process = Process()
            process.executableURL = executableURL
            process.arguments = [
                "-p",
                "--no-session-persistence",
                "--permission-mode",
                "dontAsk",
                "--output-format",
                "text",
            ]

            process.environment = ClaudeBinaryResolver.augmentedEnvironment()

            let stdinPipe = Pipe()
            let stdoutPipe = Pipe()
            let stderrPipe = Pipe()
            process.standardInput = stdinPipe
            process.standardOutput = stdoutPipe
            process.standardError = stderrPipe

            let timeoutWorkItem = DispatchWorkItem {
                if process.isRunning {
                    process.terminate()
                    continuation.resume(throwing: SummaryServiceError.timedOut)
                }
            }
            DispatchQueue.global().asyncAfter(deadline: .now() + summaryTimeoutSeconds, execute: timeoutWorkItem)

            process.terminationHandler = { proc in
                timeoutWorkItem.cancel()
                let stdoutData = stdoutPipe.fileHandleForReading.readDataToEndOfFile()
                let stderrData = stderrPipe.fileHandleForReading.readDataToEndOfFile()
                let stdout = String(data: stdoutData, encoding: .utf8)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
                let stderr = String(data: stderrData, encoding: .utf8)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""

                if proc.terminationStatus != 0 {
                    continuation.resume(throwing: SummaryServiceError.cliFailed(stderr.isEmpty ? "Claude CLI exited with code \(proc.terminationStatus)" : stderr))
                    return
                }

                if stdout.isEmpty {
                    continuation.resume(throwing: SummaryServiceError.cliFailed(
                        "Claude CLI returned empty output. Run `claude -p \"hello\"` in Terminal to verify your CLI setup."
                    ))
                    return
                }

                continuation.resume(returning: stdout)
            }

            do {
                try process.run()
                if let promptData = prompt.data(using: .utf8) {
                    stdinPipe.fileHandleForWriting.write(promptData)
                }
                try stdinPipe.fileHandleForWriting.close()
            } catch {
                timeoutWorkItem.cancel()
                continuation.resume(throwing: SummaryServiceError.cliFailed("Failed to run Claude CLI (\(config.claudeBinary)): \(error.localizedDescription)"))
            }
        }
    }
}

enum SummaryServiceError: LocalizedError {
    case timedOut
    case cliFailed(String)

    var errorDescription: String? {
        switch self {
        case .timedOut:
            return "Claude CLI timed out after 120 seconds"
        case .cliFailed(let message):
            return message
        }
    }
}
